### ISSUE TITLE
*feature request* Suggestions about super_lm 

### ISSUE THREAD

**Issue by AranKomat** [tensorflow/tensor2tensor#758]
As I've been working on LM and (un-)conditional text generation with T-D, I'd like to make a few suggestions for super_lm. 

- You may want to try adaptive softmax as in [1] and/or mixture of softmaxes. The latter slows down the training but remarkably improves the perplexity especially for a large dataset like 1 Billion Words. Even mixup may also be applicable. 

- If your dataset requires to keep track of over a thousand of steps, then it's better to fortify T-D with auxiliary loss as in [2], as the performance degradation is visible after that point, not to mention that T-D may be replaced with T-DMCA or something like that to save memory. 

-  I compared MLE T-D with SOTA text generation algorithms such as LeakGAN and CoT [3] with WMT News 2017 dataset (~200,000 sentences, ~30 words each). T-D achieved by far the best NLL yet, without beam search, slightly lower BLEU-n than LeakGAN. 

- If you want to evaluate generated long sequences, I recommend MMD with Word Mover's Distance over BLEU-n, the latter of which is time-consuming if the length exceeds 50. It must be very fun to generate music, make a conversation agent and try various other seq2seq stuffs with T-D. 

[1] An Analysis of Neural Language Modeling at Multiple Scales
[2] Learning Longer-term Dependencies in RNNs with Auxiliary Losses
[3] CoT: Cooperative Training for Generative Modeling
